drivers/i2s/i2s_cavs.c

```diff
@@ -0,0 +1,791 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file
+ * @brief I2S bus (SSP) driver for Intel CAVS.
+ *
+ * Limitations:
+ * - DMA is used in simple single block transfer mode (with linked list
+ *   enabled) and "interrupt on full transfer completion" mode.
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <misc/__assert.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <dma.h>
+#include <i2s.h>
+#include <soc.h>
+#include "i2s_cavs.h"
+
+#define SYS_LOG_DOMAIN "dev/i2s_cavs"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_I2S_LEVEL
+#include <logging/sys_log.h>
+
+#ifdef CONFIG_DCACHE_WRITEBACK
+#define DCACHE_INVALIDATE(addr, size) \
+		{ dcache_invalidate_region(addr, size); }
+#define DCACHE_CLEAN(addr, size) \
+		{ dcache_writeback_region(addr, size); }
+#else
+#define DCACHE_INVALIDATE(addr, size) \
+		do { } while (0)
+
+#define DCACHE_CLEAN(addr, size) \
+		do { } while (0)
+#endif
+
+#define CAVS_SSP_WORD_SIZE_BITS_MIN    2
+#define CAVS_SSP_WORD_SIZE_BITS_MAX   32
+#define CAVS_SSP_WORD_PER_FRAME_MIN    1
+#define CAVS_SSP_WORD_PER_FRAME_MAX   16
+
+struct queue_item {
+	void *mem_block;
+	size_t size;
+};
+
+/* Minimal ring buffer implementation:
+ * buf - holds the pointer to Queue items
+ * len - Max number of Queue items that can be referenced
+ * head - current write index number
+ * tail - current read index number
+ */
+struct ring_buf {
+	struct queue_item *buf;
+	u16_t len;
+	u16_t head;
+	u16_t tail;
+};
+
+/* This indicates the Tx/Rx stream. Most members of the stream are
+ * self-explanatory except for sem.
+ * sem - This is initialized to CONFIG_I2S_CAVS_TX_BLOCK_COUNT. If at all
+ * mem_block_queue gets filled with the MAX blocks configured, this semaphore
+ * ensures nothing gets written into the mem_block_queue before a slot gets
+ * freed (which happens when a block gets read out).
+ */
+struct stream {
+	s32_t state;
+	struct k_sem sem;
+	u32_t dma_channel;
+	struct dma_config dma_cfg;
+	struct i2s_config cfg;
+	struct ring_buf mem_block_queue;
+	void *mem_block;
+	bool last_block;
```

Comment:
With the embedded structs here, will having a boolean in the middle create a unnecessary hole due to alignment?

Reply:
Well, not sure if a hole is created here. What is the size of bool in zephyr?